### PR TITLE
Extended proof chain algorithms

### PR DESCRIPTION
**Pull request by Wind4Greg**
This PR addresses issues https://github.com/w3c/vc-data-integrity/issues/87 and https://github.com/w3c/vc-data-integrity/issues/120.  The proof chain *addition* and *verification* algorithms are extended to allow the case where `previousProof` is an unordered  list, i.e., the validity of a proof (signature) may dependent on two or more independent signatures.

In addition the verification algorithm warns implementers on the dangers of **cycles** that could occur in a malicious proof chain and suggests on simple method of detection via tracking visited proofs. Note that cycle detection is a fairly well established area in computer science and graph theory so kept this discussion fairly short.

***
<a href="https://pr-preview.s3.amazonaws.com/Wind4Greg/vc-data-integrity/pull/145.html" title="Last updated on Aug 14, 2023, 4:57 PM UTC (0d2a850)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/145/3b3e64a...Wind4Greg:0d2a850.html" title="Last updated on Aug 14, 2023, 4:57 PM UTC (0d2a850)">Diff</a>

### CODE REVIEW

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
An OPTIONAL string value or unordered list of string values. Each value 
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
attribute is a string, check whether 
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
If not, a `PROOF_GENERATION_ERROR` MUST be raised. If the 
<code>previousProof</code> attribute is an array, check that each element of that
array is a string and that an element of the <var>proof_list</var> has a 
matching <code>id</code> attribute. If not, a `PROOF_GENERATION_ERROR` MUST be 
```

File: index.html
Comment (TallTed) [inline index.html]: this feels like a repeat suggestion...
```suggestion
In a proof set or proof chain, a <a>secured data document</a> has a
```

File: index.html
Comment (TallTed) [inline index.html]: The change above reverts a prior suggestion from me. I'm making a slightly different suggestion here.
```suggestion
For each <var>proof</var> in <var>proof_list</var>, run steps 2 through 11 of the algorithm in
section <a href="verify-proof"></a>; if no exceptions are raised, associate
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
<code>previousProof</code> attribute, modify the associated 
<var>isProofVerified</var> by iteratively checking that the proof identified by
the <code>id</code> value in <var>previousProof</var>, along with any 
<code>previousProofs</code> up the chain, is valid. If <code>previousProof</code> is an
array, then all previous proof proof chains identified by the array elements 
MUST be valid for the current proof to be considered valid. Cycle detection, 
such as keeping a list of previously visited proofs, MUST be used to guard 
against malicious proofs. If a cycle is detected in a proof chain, 
then a `PROOF_CHAIN_CYCLE_ERROR` MUST be raised.
```

**Review by TallTed — CHANGES_REQUESTED** [review]
I think there's been a collision of PRs, possibly across branches. Some of these requested changes are re-requests of changes that were applied previously. I mention this as it means that all of us have to be doubly careful when re-reviewing areas that we've already reviewed.

File: index.html
Comment (dmitrizagidulin) [inline index.html]:
```suggestion
Return the <var>proof_list</var> along with each proof's associated 
```
The original phrasing with the apostrophe, is correct.

**Review by dmitrizagidulin — CHANGES_REQUESTED** [review]

File: index.html
Comment (dlongley) [inline index.html]:
```suggestion
the <code>id</code> value in <var>previousProof</var>, along with every 
<code>previousProof</code> up the chain, is valid. If <code>previousProof</code> is an
```

File: index.html
Comment (dlongley) [inline index.html]:
```suggestion
array, then all previous proofs identified by the array elements 
```

**Review by dlongley — APPROVED** [review]

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
current proof is processed. If an unordered list, all referenced 
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
array is a string which matches the <code>id</code> attribute of an element
of the <var>proof_list</var>. If not, a `PROOF_GENERATION_ERROR` MUST be 
```

**Review by TallTed — CHANGES_REQUESTED** [review]

File: index.html
Comment (msporny) [inline index.html]: @Wind4Greg, please set your editor to "strip trailing whitespace". This PR is re-introducing trailing spacing characters that were removed in another PR. We do not use trailing whitespace on W3C VCWG specifications.

File: index.html
Comment (msporny) [inline index.html]:
```suggestion
<var>allProofs</var> be an empty list. If <var>proof</var> is a list copy all 
```

All algorithms use camelCase. Don't care much about the name, but at least camelCase it, please. Same suggestion for all `proof_list` values below.

File: index.html
Comment (msporny) [inline index.html]: Strip trailing whitespace for these lines (and all lines below that have trailing whitespace). It makes it harder to read PRs because it makes it seem like more has changed in the PR than actually has changed.

File: index.html
Comment (msporny) [inline index.html]:
```suggestion
against proof cycles. If a cycle is detected in a proof chain, 
```

Not all cycles might be malicious... :)

**Review by msporny — CHANGES_REQUESTED** [review]
Overall algorithm looks good at first glance. 

Changes requested are: strip trailing whitespace, use camelCase variable names, don't presume maliciousness in proof cycles.

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
<var>allProofs</var> be an empty list. If <var>proof</var> is a list, copy all 
```

File: index.html
Comment (TallTed) [inline index.html]: Given the change above, of `proof_list` to `allProofs`, shouldn't that be carried through the document? As --
```suggestion
the elements of <var>proof</var> to <var>allProofs</var>. If <var>proof</var> 
is an object add a copy of that object to <var>allProofs</var>.
```
(There are more instances to change, below.)

**Comment by msporny** [comment]
@Wind4Greg ping on this PR -- waiting on you to rebase, strip trailing spaces, fix merge conflicts, etc. before we can merge this. I tried to rebase and resolve the merge conflicts, but after 3 attempts, gave up. I didn't know what you changed in this PR. I could squash-and-merge, but didn't want to lose your change history.

 LMK what you want to do here, please.

**Comment by Wind4Greg** [comment]
@msporny Can get to this Monday. If that is too late proceed.

**Comment by Wind4Greg** [comment]
I think I hit the wrong button and closed/merged this when I was just trying to clean up trailing spaces and merge mess.

**Pull request merged**
